### ISSUE TITLE
Rule proposal: Add option to `no-children-prop` to allow limited children props on self-closing tags?

### ISSUE THREAD

**Issue by vdh** [yannickcr/eslint-plugin-react#873]
Would it be possible to extend `no-children-prop` with an opt-in option to allow simple values on self-closing tags, e.g. `<div children="foo" />`? It would still disallow components or arrays, and disallow the prop on tags that are not self-closing.

I put together some mocked documentation of how an `allowSelfClosing` option might work:

Examples of **incorrect** code for this rule:
```js
/* eslint react/no-children-prop: ["error"] */
// Don't allow children in props
var jsx1 = <div children="foo" />;
var js1  = React.createElement("div", { children: "foo" });
var jsx2 = <MyComponent children="foo" />;
var js2  = React.createElement(MyComponent, { children: "foo" });

// Don't allow components or arrays
var jsx3 = <MyComponent children={<AnotherComponent />} />;
var jsx4 = <MyComponent children={["foo", "bar"]} />;
```

Examples of **correct** code for this rule:
```js
/* eslint react/no-children-prop: ["error"] */
var jsx1 = <div>foo</div>;
var js1  = React.createElement("div", null, "foo");

var jsx2 = <MyComponent>foo</MyComponent>;
var js2  = React.createElement(MyComponent, null, "foo");

var jsx3 = (
  <MyComponent>
    <span>foo</span>
    <span>bar</span>
  </MyComponent>
);
var js4  = React.createElement("div", "foo", "bar");
```

When `allowSelfClosing` is set to `true` the following patterns are no longer considered as warnings:
```js
/* eslint react/no-children-prop: ["error", { "allowSelfClosing": true }] */
var jsx1 = <div children="foo" />;
var js1  = React.createElement("div", { children: "foo" });
var jsx2 = <MyComponent children="foo" />;
var js2  = React.createElement(MyComponent, { children: "foo" });
```

Examples of **incorrect** code for this rule when `allowSelfClosing` is set to `true`:
```js
/* eslint react/no-children-prop: ["error", { "allowSelfClosing": true }] */

// Don't allow mixing children prop with children
var jsx1 = <div children="foo">bar</div>;
var js1  = React.createElement("div", { children: "foo" }, "bar");
var jsx2 = <MyComponent children="foo">bar</MyComponent>;
var js2  = React.createElement(MyComponent, { children: "foo" }, "bar");

// Don't allow components or arrays in children prop
var jsx3 = <MyComponent children={<AnotherComponent />} />;
var jsx4 = <MyComponent children={["Child 1", "Child 2"]} />;
```

**Comment by ljharb** [comment]
What's the concrete benefit of this? The rationale for the rule is specifically that you should *never*, under any circumstances, directly use the "children" prop, and should only ever pass children as the third argument of `React.createElement`, or as jsx descendants.

**Comment by vdh** [comment]
@ljharb When a leaf node has a lot of props, especially when nesting limits the available horizontal space, the opening tag is often broken onto multiple lines:
```jsx
<MyComponent
  foo="bar"
  baz="qux"
  otherPropWithLongName={10}
>Label Text</MyComponent>
```
Which could be easier to read when written with the `children` prop:
```jsx
<MyComponent
  foo="bar"
  baz="qux"
  otherPropWithLongName={10}
  children="Label Text"
/>
```

**Comment by ljharb** [comment]
@vdh You'd just do:
```jsx
<MyComponent
  foo="bar"
  baz="qux"
  otherPropWithLongName={10}
>
  Label Text
</MyComponent>
```
since React collapses all the whitespace anyways. If you didn't want the single leading and trailing space, you could do something like:
```jsx
<MyComponent
  foo="bar"
  baz="qux"
  otherPropWithLongName={10}
>{
  'Label Text'
}</MyComponent>
```
or you'd just disable the rule for a use case where the exact position of whitespaces mattered.

**Comment by tran-simon** [comment]
It is useful if you need to override the children prop

`
const Comp = (ButtonProps)=>{
  return <Button children='Default Text' {...ButtonProps}/>
}

const Comp1 = ()=>{
  return <Comp>AlternativeText</Comp>
}

const Comp2 = ()=>{
  return <Comp/>
}

`

**Comment by ljharb** [comment]
@tran-simon that’s what .defaultProps is for.
